### PR TITLE
Have inheritance-margin hold onto data that does *not* point back at Solution instances.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: src/EditorFeatures/Test/InheritanceMargin/InheritanceMarginTests.cs [hunk 1]
```diff
@@ -81,11 +81,11 @@ private static Task VerifyNoItemForDocumentAsync(string markup, string languageN
 
             for (var i = 0; i < sortedActualItems.Length; i++)
             {
-                VerifyInheritanceMember(testWorkspace, sortedExpectedItems[i], sortedActualItems[i]);
+                await VerifyInheritanceMember(testWorkspace, sortedExpectedItems[i], sortedActualItems[i]);
             }
         }
 
-        private static void VerifyInheritanceMember(TestWorkspace testWorkspace, TestInheritanceMemberItem expectedItem, InheritanceMarginItem actualItem)
+        private static async Task VerifyInheritanceMember(TestWorkspace testWorkspace, TestInheritanceMemberItem expectedItem, InheritanceMarginItem actualItem)
```
Comment (JoeRobich) [inline src/EditorFeatures/Test/InheritanceMargin/InheritanceMarginTests.cs:88]: Should `Task` returning functions have the `Async` suffix?

Comment (sharwell) [reply src/EditorFeatures/Test/InheritanceMargin/InheritanceMarginTests.cs:88]: They should yes. Need to update https://github.com/dotnet/roslyn/pull/51812.

Comment (CyrusNajmabadi) [reply src/EditorFeatures/Test/InheritanceMargin/InheritanceMarginTests.cs:88]: in tests we're all over the map.

**Comment by CyrusNajmabadi** [comment]
@sharwell @JoeRobich can you ptal?  tahnks!

File: src/Features/Core/Portable/FindUsages/DefinitionItem.DetachedDefinitionItem.cs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+            public async Task<DefaultDefinitionItem?> TryRehydrateAsync(CancellationToken cancellationToken)
+            {
+                using var converted = TemporaryArray<DocumentSpan>.Empty;
+                foreach (var ss in SourceSpans)
+                {
+                    var documentSpan = await ss.TryRehydrateAsync(cancellationToken).ConfigureAwait(false);
+                    if (documentSpan == null)
+                        return null;
+                }
```
Comment (JoeRobich) [inline src/Features/Core/Portable/FindUsages/DefinitionItem.DetachedDefinitionItem.cs:54]: Should `documentSpan` be added to `converted`?

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/FindUsages/DefinitionItem.DetachedDefinitionItem.cs:54]: yes :D

File: src/Features/Core/Portable/FindUsages/DefinitionItem.DetachedDefinitionItem.cs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+            }
+
+            public async Task<DefaultDefinitionItem?> TryRehydrateAsync(CancellationToken cancellationToken)
+            {
+                using var converted = TemporaryArray<DocumentSpan>.Empty;
+                foreach (var ss in SourceSpans)
+                {
+                    var documentSpan = await ss.TryRehydrateAsync(cancellationToken).ConfigureAwait(false);
+                    if (documentSpan == null)
+                        return null;
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/FindUsages/DefinitionItem.DetachedDefinitionItem.cs:53]:
```suggestion
                        return null;
                        
                    converted.Add(documentSpan.Value);
```

**Review by JoeRobich — APPROVED** [review]
I could see there being tests around the Detach()/ReHydrate(), if only because of the omission we caught.

**Comment by CyrusNajmabadi** [comment]
> I could see there being tests around the Detach()/ReHydrate(), if only because of the omission we caught.

@Cosifne do we have integration tests for inheritance margin?  if not, can we file an issue to add them asap?

**Comment by Cosifne** [comment]
> > I could see there being tests around the Detach()/ReHydrate(), if only because of the omission we caught.
> 
> @Cosifne do we have integration tests for inheritance margin? if not, can we file an issue to add them asap?

There are only unit tests in for inheritance Margin.
Sure I can file a issue for that

**Comment by CyrusNajmabadi** [comment]
> There are only unit tests in for inheritance Margin.
Sure I can file a issue for that

Thanks.  It's always good to have at least one integration test to totally validate the end to end, (esp. with custom UI).

**Review by Cosifne — APPROVED** [review]
I don't see any problem in the Inheritance Margin part : )

File: src/VisualStudio/Core/Def/Implementation/InheritanceMargin/MarginGlyph/InheritanceMargin.xaml.cs [hunk 4]
```diff
@@ -65,22 +72,33 @@ private void TargetMenuItem_OnClick(object sender, RoutedEventArgs e)
...
 
+        private async Task TargetMenuItem_OnClickAsync(TargetMenuItemViewModel viewModel)
+        {
+            using var context = _operationExecutor.BeginExecute(
+                title: EditorFeaturesResources.Navigating,
+                defaultDescription: string.Format(ServicesVSResources.Navigate_to_0, viewModel.DisplayContent),
+                allowCancellation: true,
+                showProgress: false);
+
+            var rehydrated = await viewModel.DefinitionItem.TryRehydrateAsync(context.UserCancellationToken).ConfigureAwait(false);
```
Comment (Cosifne) [inline src/VisualStudio/Core/Def/Implementation/InheritanceMargin/MarginGlyph/InheritanceMargin.xaml.cs:89]: Does GoToDefinitionHelpers.TryGoToDefinition below needs to be called by UI thread?

Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Implementation/InheritanceMargin/MarginGlyph/InheritanceMargin.xaml.cs]: switched here to be entirely async.

File: src/EditorFeatures/Core/GoToDefinition/GoToDefinitionHelpers.cs [hunk 5]
```diff
@@ -122,28 +122,9 @@ internal static class GoToDefinitionHelpers
             IStreamingFindUsagesPresenter streamingPresenter,
             CancellationToken cancellationToken)
         {
-            if (definitions.IsDefaultOrEmpty)
-                return false;
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/GoToDefinition/GoToDefinitionHelpers.cs:126]: this moved inside TryNavigateToOrPresentItemsAsync

**Comment by 333fred** [comment]
@CyrusNajmabadi it looks like there's some merge commits in here that are likely bringing in code that was originally reviewed (as it's now requesting compiler reviews), so I disabled automerge so you can make sure everything is kosher.

**Pull request merged**